### ISSUE TITLE
Code refactoring - an Agile approach

### ISSUE THREAD

**Issue by butor** [butor/blackbird#234]
@sbarale
@greatwolf 

As we know, @sbarale is working on refactoring the Blackbird codebase. As he stated, the goal is to have a more extensible platform and less prone to errors when adding/modifying things, by having a separation of concerns in place and having the trading logic to be _modular_ as well as each exchange’s implementation.

I think that's definitively a great idea. I'm not a C++ expert but the refactoring @sbarale has started to implement seems to be going in the right direction ([see his code here](https://github.com/sbarale/blackbird/tree/tidying-up)).

The primary concern of @greatwolf is that the whole refactoring is implemented in one PR (see PR #231), and his concern makes sense.

I was discussing with @sbarale this morning, to see if we can have an Agile approach and implement his refactoring in an iterative, Agile way. Small commits at a time, and we can comment, test and validate them iteratively. I know it's difficult when doing a large refactoring, but I think it's the best option.

The good news is that this refactoring will only have a limited impact on the "business" features of Blackbird, so the regression testing should be easier to do.

I have created this issue to discuss anything about that refactoring.

@greatwolf Let us know what you think about this approach. Thanks

**Comment by sbarale** [comment]
Thanks @butor for the intro. @greatwolf as I mentioned previously, it's been a while I don't work in C/C++ so i'm rusty at best.. feel free to suggest any changes you see fit.

In my years of development, one of the things I dislike the most are migrations. For this reason, if there's not a lot of open tasks and given the fact that the changes proposed make a 180 degree change on how the code is organized, I think it would be best to do it in one swoop (NOTE: I would never do this if there is a lot of outstanding tasks).

With the changes I've implemented, anyone currently working on the code will be able to pickup and continue working with very few considerations.

IOW, if people were working on an exchange implementation, now they just need to know that instead of a namespace, they'll be using classes extending the AbstractExchange interface.

The next step would be to isolate the logic into its own module so then it can be implemented via a strategy pattern (or something like that) thus allowing extensibility on the trading strategies allowed by the system.

Furthermore, I've taken a stab at renaming the former Bitcoin object to something more generic (renamed it Symbol for now). I think, these symbols should (could) contain references to the exchanges that can handle them and create 'trading pairs' with them thus making it easier to do multi-symbols trading and/or different strategies (i.e. between two-exchanges or three-exchanges with symbols (coin) combinations.

**Comment by sbarale** [comment]
NOTE: One more thing I forgot... whoever is good detecting leaks, please feel free to pitch-in for (as I mentioned) I'm very very rusty and I'm sure things are falling out of my scope.

**Comment by butor** [comment]
Some interesting consideration in [this new #221 comment](https://github.com/butor/blackbird/issues/221#issuecomment-312721216). Thanks @sluther

**Comment by sluther** [comment]
I think this is a fantastic idea and I'd like to start contributing when I have some time, which will probably be this coming weekend. I'd be happy to help by designing architecture or writing implementation, or some mixture of both.

Is there some kind of Slack (or IRC) channel for this?

**Comment by butor** [comment]
@sbarale we have a dedicated chat room [here](https://gitter.im/blackbird_bitcoin_arbitrage/Lobby?source=orgpage)

**Comment by uhaile** [comment]
I have just started looking at Blackbird, and am already confused. There is the original source, which is being modified, and there is the refactored one, by @sbarale.
Which one is the future? Are both going to exist in parallel? Until when, or at some stage the original one will disappear?

I think that working on one "product" and making incremental changes, even if they are large sometimes, would speed up the development, instead of having two separate tracks.

On a separate issue, perhaps saving the exchanges and currency data in sqlite tables would be more effective than having the relationships stored in config and .h files.

And finally, I find that the graph in the readme for Blackbird, explaining when arbitrage opportunities arise is very helpful. Does anybody know if there is an implementation of this graphing that would run at the same time as Blackbird is running.

**Comment by greatwolf** [comment]
I agree with the points raised by @uhaile. Particular about saving exch and currency data into sql tables -- this is congruent with the idea of keeping data and programming logic separate.

While I'm not opposed to using an OO approach of refactoring, there are things to be careful of. Packaging state and methods into neat classes makes it easier to reason about, it isn't necessarily the most performant -- you have to also consider what data is involved, how it's used and arranging it so its cache friendly. Another consequence of using OO, abstract classes invariably leads to virtual methods which means a vtable for every exchange implementation + an extra vptr member for every instantiation, extra indirection for virtual calls etc. I actually did started refactoring with this approach in my [refactor_loading branch](https://github.com/greatwolf/blackbird/commit/a061e6d2f1f95ec29960c83be03c4c35ff0f7828) but didn't really finish it. You'll notice a lot of similarities w/ @sbarale current refactor attempt.

One other aspect to consider, many of the properties that's being checked for at runtime are in fact knowable at compile-time. Like for example, whether an exchange is implemented or not, whether it can margin short etc. Refactoring those into compile-time type traits or policies using a combination of TMP and SFINAE techniques would allow removing many of those runtime `if` checks which means better performance.

Using a [reactivex](http://reactivex.io/) approach might be better since it forces a more data oriented design. This is basically like "spreadsheet programming" but instead of using cells to connect dataflow relationships you use streams. One possible design is to have the exchange act as a source for one end the pipe and the other end of the pipe produces actionable signals, like eg. open long on one exchange and margin short on second or there's a signal to close the arbitrage etc. Incidentally this kind of design makes it easier to replace the REST interface with a websocket one by just replacing the data source(will need another library for this since cURL doesn't support websocks afaik).

**Comment by sbarale** [comment]
@greatwolf thx for your 2 satoshis (man I love that line... lol).
While OO could make it less performant, I don't think it's a matter of concern here, is it? since we're talking about the most performant language there is outthere for the most part. All of the other trading codes I've seen are in other languages, even interpreted ones.

I think it all depends on what the ultimate goal is... In my opinion going for OO it provides a more flexible, decoupled way of implementing it which in the end would (should) be less prone to errors when the time for expansion/modification.

With that said, I'm not a master in architecting... I successfully built several big scale project in my past but which achieved their goal as far as a compromise between stability, scalability and performance but probably my approach is wrong and there are far betters ways to go about it, which I'm all ears; I love to learn new things.

Furthermore we can agree to disagree and each fork our own way... that's fine too... my involvement in this project was precisely b/c I thought I could bring something to the table which would lead to faster development given that, in the end, bit spread arbitrage opportunities are a race against time.

With that said, @greatwolf what do you suggest we do? When I talked to @butor he mentioned you were the lead when it came to the software/language related decisions and I'll respect that.

**Comment by sluther** [comment]
I understand there are potential performance implications with the approach I proposed, but I think my approach will also make it more useful and usable, and I think the minor performance impact is worth the tradeoff. Going the OO route also opens up the door for making testable, scalable, easy to understand code. All of these things, in my opinion, are worth the performance hit of using OO - especially since we can always come up with a way to determine what currency pairs are valid and cache it somewhere. We could even cache it in SQL!

**Comment by antony** [comment]
A critical part of why Agile works is tests. I don't see any mention here about testing, unit, integration etc, but this would be a great way to prevent regressions as you extend blackbird's functionality.

Sadly, I know nothing about testing in C++, so I can't help other than to suggest that tests are added as part of the refactoring, and some sort of CI solution (travis, for example) to run them on every commit.

**Comment by sluther** [comment]
@antony I mentioned tests in my comment right above yours. And of course, I completely agree with you. I don't know anything about testing in C++, that doesn't mean I'm not willing to learn. In fact, I'd be more than happy to have my contribution be mainly unit tests - as it happens, I really enjoy writing unit and integration tests, both before and after implementation - even if that implementation is someone else's.

**Comment by rluijk** [comment]
My two cents. Make small increments so we can all work together! It is a community and opensource projects move the fastest when people CAN contribute instead of waiting. It makes more sense to do it together in baby steps and getting the best from the best.
